File: max_subarray.java
//TC: O(n)
//SC: O(1)

public class max_subarray {
    class Solution {
        public int maxSubArray(int[] nums) {
            int largestSum = nums[0];
            int currentSum = nums[0];
            
            for (int i = 1; i < nums.length; i++) {
                if (currentSum + nums[i] > nums[i]) {
                    currentSum = currentSum + nums[i];
                } else {
                    currentSum = nums[i];
                }
                
                if (currentSum > largestSum) {
                    largestSum = currentSum;
                }
            }
            
            return largestSum;
        }
    }
}

File: maxsubarray.swift
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
    func maxSubArray(_ nums: [Int]) -> Int {
        if nums == nil ||  nums.count == 0 {
            return 0
        }
        var maxSum = Int.min
        var sum = 0
        for i in  0..<nums.count {
            sum = sum + nums[i]
            if sum < nums[i] {
                sum = nums[i]
            }
            if sum > maxSum {
                maxSum = sum
            }
        }
        return maxSum
    }